src/main/java/com/google/devtools/build/lib/bazel/bzlmod/VendorManager.java

```diff
@@ -57,25 +57,36 @@ public void vendorRepos(Path externalRepoRoot, ImmutableList<RepositoryName> rep
     }
 
     for (RepositoryName repo : reposToVendor) {
-      // Only re-vendor the repository if it is not up-to-date.
-      if (!isRepoUpToDate(repo, externalRepoRoot)) {
-        Path markerUnderVendor = vendorDirectory.getChild(repo.getMarkerFileName());
-        Path repoUnderVendor = vendorDirectory.getRelative(repo.getName());
-
-        // 1. Clean up existing marker file and repo vendor directory
-        markerUnderVendor.delete();
-        repoUnderVendor.deleteTree();
-        repoUnderVendor.createDirectory();
-
-        // 2. Copy over the repo source.
-        FileSystemUtils.copyTreesBelow(
-            externalRepoRoot.getRelative(repo.getName()), repoUnderVendor, Symlinks.NOFOLLOW);
+      Path repoUnderExternal = externalRepoRoot.getChild(repo.getName());
+      Path repoUnderVendor = vendorDirectory.getRelative(repo.getName());
+      // This could happen when running the vendor command twice without changing anything.
+      if (repoUnderExternal.isSymbolicLink() && repoUnderExternal.resolveSymbolicLinks().compareTo(repoUnderVendor) == 0) {
+        continue;
+      }
 
-        // 3. Copy the marker file atomically
-        Path tMarker = vendorDirectory.getChild(repo.getMarkerFileName() + ".tmp");
-        FileSystemUtils.copyFile(externalRepoRoot.getChild(repo.getMarkerFileName()), tMarker);
-        tMarker.renameTo(markerUnderVendor);
+      // At this point, the repo should exist under external dir, but check if the vendor src is already up-to-date.
+      Path markerUnderExternal = externalRepoRoot.getChild(repo.getMarkerFileName());
+      Path markerUnderVendor = vendorDirectory.getChild(repo.getMarkerFileName());
+      if (isRepoUpToDate(markerUnderVendor, markerUnderExternal)) {
+        continue;
       }
+
+      // Actually vendor the repo:
+      // 1. Clean up existing marker file and vendor dir.
+      markerUnderVendor.delete();
+      repoUnderVendor.deleteTree();
+      repoUnderVendor.createDirectory();
+      // 2. Move the marker file to a temporary one under vendor dir.
+      Path tMarker = vendorDirectory.getChild(repo.getMarkerFileName() + ".tmp");
+      FileSystemUtils.moveFile(markerUnderExternal, tMarker);
+      // 3. Move the external repo to vendor dir. It's fine if this step fails or is interrupted, because the marker
+      // file under external is gone anyway.
+      FileSystemUtils.moveTreesBelow(repoUnderExternal, repoUnderVendor);
+      // 4. Rename to temporary marker file after the move is done.
+      tMarker.renameTo(markerUnderVendor);
+      // 5. Leave a symlink in external dir.
+      repoUnderExternal.deleteTree();
```

Comment:
is this necessary? we just moved it.

Reply:
I'll only move content below the dir without deleting it.